src/coreclr/jit/morph.cpp

```diff
@@ -323,6 +323,222 @@ GenTree* Compiler::fgMorphExpandCast(GenTreeCast* tree)
+                oper = gtNewSimdCndSelNode(TYP_SIMD16, oper, maxValDup, newTree, destFieldType, 16);
+                oper = gtNewSimdHWIntrinsicNode(dstType, oper, NI_Vector128_ToScalar, destFieldType, 16);
+                return fgMorphTree(oper);
+            }
+            else if (dstType == TYP_UINT && compOpportunisticallyDependsOn(InstructionSet_SSE41))
+            {
+                if (srcType == TYP_FLOAT)
+                {
+                    oper      = gtNewCastNode(TYP_DOUBLE, oper, false, TYP_DOUBLE);
+                    srcType   = TYP_DOUBLE;
+                    fieldType = CORINFO_TYPE_DOUBLE;
+                }
+                ssize_t     actualMaxVal  = UINT32_MAX;
+                CorInfoType destFieldType = CORINFO_TYPE_LONG;
+
+                oper               = gtNewSimdCreateBroadcastNode(TYP_SIMD16, oper, fieldType, 16);
+                GenTree* op1Clone1 = fgMakeMultiUse(&oper);
+                GenTree* op1Clone2 = fgMakeMultiUse(&oper);
+                GenTree* op1Clone3 = fgMakeMultiUse(&oper);
+                GenTree* op1Clone4 = fgMakeMultiUse(&oper);
+
+                // get the max/min value vector
+                GenTree* minVal    = gtNewDconNodeD(static_cast<double>(0));
+                minVal             = gtNewSimdCreateBroadcastNode(TYP_SIMD16, minVal, fieldType, 16);
+                GenTree* maxVal    = gtNewDconNodeD(static_cast<double>(actualMaxVal));
+                maxVal             = gtNewSimdCreateBroadcastNode(TYP_SIMD16, maxVal, fieldType, 16);
+                GenTree* maxValDup = fgMakeMultiUse(&maxVal);
+
+                // check NaN
+                GenTree* mask1 = gtNewSimdCmpOpNode(GT_EQ, TYP_SIMD16, oper, op1Clone1, fieldType, 16);
+                // check negative
+                GenTree* mask2 = gtNewSimdCmpOpNode(GT_GE, TYP_SIMD16, op1Clone2, minVal, fieldType, 16);
+
+                // and mask
+                GenTree* mask12 = gtNewSimdBinOpNode(GT_AND, TYP_SIMD16, mask1, mask2, fieldType, 16);
+
+                // inp = inp & mask
+                GenTree* saturated_val   = gtNewSimdBinOpNode(GT_AND, TYP_SIMD16, op1Clone3, mask12, fieldType, 16);
+                GenTree* saturatedValDup = fgMakeMultiUse(&saturated_val);
+
+                // compare with max value of uint
+                GenTree* mask3 = gtNewSimdCmpOpNode(GT_GE, TYP_SIMD16, saturated_val, maxVal, fieldType, 16);
+
+                // Select based on mask3
+                saturated_val = gtNewSimdCndSelNode(TYP_SIMD16, mask3, maxValDup, saturatedValDup, fieldType, 16);
+
+                // scalar
+                GenTree* val = gtNewSimdHWIntrinsicNode(srcType, saturated_val, NI_Vector128_ToScalar, fieldType, 16);
+
+                // cast it
+                tree = gtNewCastNode(TYP_INT, val, false, dstType);
+                tree->SetSaturatedConversion();
+                return fgMorphTree(tree);
```

Comment:
This is just going to `nop` right? I think its the thing that is "requiring" the `SetSaturatedConversion` and which we could avoid by simply marking `tree` as morphed in debug (as other places in morph already do).

Reply:
few days back, I tried marking all saturated nodes as morphed but they started hitting ```Node is already Morphed``` assert in morph. I will investigate further and get back.